worlds/tww/Macros.py

```diff
@@ -0,0 +1,1544 @@
+
+def has_mirror_shield(state: CollectionState, player: int) -> bool:
+    return state.has("Progressive Shield", player, 2)
+
+
+def has_heros_bow(state: CollectionState, player: int) -> bool:
+    return state.has("Progressive Bow", player, 1)
+
+
+def has_fire_arrows(state: CollectionState, player: int) -> bool:
+    return state.has("Progressive Bow", player, 2) and has_magic_meter(state, player)
+
+
+def has_ice_arrows(state: CollectionState, player: int) -> bool:
+    return state.has("Progressive Bow", player, 2) and has_magic_meter(state, player)
+
+
+def has_light_arrows(state: CollectionState, player: int) -> bool:
+    return state.has("Progressive Bow", player, 3) and has_magic_meter(state, player)
+
+
+def has_any_wallet_upgrade(state: CollectionState, player: int) -> bool:
+    return state.has("Wallet Capacity Upgrade", player, 1)
+
+
+def has_picto_box(state: CollectionState, player: int) -> bool:
+    return state.has("Progressive Picto Box", player, 1)
+
+
+def has_deluxe_picto_box(state: CollectionState, player: int) -> bool:
+    return state.has("Progressive Picto Box", player, 2)
+
+
+def has_60_bomb_bomb_bag(state: CollectionState, player: int) -> bool:
+    return state.has("Bomb Bag Capacity Upgrade", player, 1)
+
+
+def has_99_bomb_bomb_bag(state: CollectionState, player: int) -> bool:
+    return state.has("Bomb Bag Capacity Upgrade", player, 2)
+
+
+def has_60_arrow_quiver(state: CollectionState, player: int) -> bool:
+    return state.has("Quiver Capacity Upgrade", player, 1)
+
+
+def has_99_arrow_quiver(state: CollectionState, player: int) -> bool:
+    return state.has("Quiver Capacity Upgrade", player, 2)
+
+
+def has_magic_meter(state: CollectionState, player: int) -> bool:
+    return state.has("Progressive Magic Meter", player, 1)
+
+
+def has_magic_meter_upgrade(state: CollectionState, player: int) -> bool:
+    return state.has("Progressive Magic Meter", player, 2)
+
+
+def has_all_8_triforce_shards(state: CollectionState, player: int) -> bool:
+    return state.has_group_unique("Shards", player, 8)
+
+
+def has_tingle_bombs(state: CollectionState, player: int) -> bool:
+    return state.has("Bombs", player) or (state._tww_tuner_logic_enabled(player) and state.has("Tingle Tuner", player))
+
+
+def can_activate_tingle_bomb_triggers_without_tingle_tuner(state: CollectionState, player: int) -> bool:
+    return True
+
+
+def can_reach_outset_island_upper_level(state: CollectionState, player: int) -> bool:
+    return can_cut_down_outset_trees(state, player) or (
+        can_fly_with_deku_leaf_outdoors(state, player) and state._tww_obscure_1(player)
+    )
+
+
+def can_access_forest_of_fairies(state: CollectionState, player: int) -> bool:
+    return can_reach_outset_island_upper_level(state, player) and can_fly_with_deku_leaf_outdoors(state, player)
+
+
+def can_reach_dragon_roost_cavern_gaping_maw(state: CollectionState, player: int) -> bool:
+    return (
+        can_access_dragon_roost_cavern(state, player)
```

Comment:
There are many rules that check whether a specific region can be accessed, but this is often redundant because the rule is only used on locations that are in that region to begin with.

In this example, `can_reach_dragon_roost_cavern_gaping_maw()` calls `can_access_dragon_roost_cavern()` which checks for being able to reach the `Dragon Roost Cavern` region. But `can_reach_dragon_roost_cavern_gaping_maw()` is only used on the `Dragon Roost Cavern - Tingle Statue Chest` and `Dragon Roost Cavern - Big Key Chest` locations, both of which are in the `Dragon Roost Cavern` region to begin with.

Reply:
Interesting, I was not aware that's how AP's algorithm works. I'll look through the rules are try to remove all the redundant checks.

Reply:
`Location.can_reach()` checks if the parent region is accessible first: https://github.com/ArchipelagoMW/Archipelago/blob/b7baaed39112d9a5e1c5b17b1914bf5205328535/BaseClasses.py#L1269-L1272